sdk/aqueduct/utils/naming.py

```diff
@@ -0,0 +1,101 @@
+
+    Operator collisions are resolved. Artifact collisions are not tolerated.
+
+    Args:
+        dag:
+            The current dag.
+        candidate_op_name:
+            The operator name that we want to check and potentially deduplicate.
+        overwrite_existing_op_name:
+            If set, any operator collisions will be ignored. Otherwise, we will bump
+            the operator name until an available one is found.
+        candidate_artifact_names:
+            The custom output artifact names the caller wants to assign to the operator
+            outputs. If not set, we will use our default artifact naming scheme. Regardless,
+            if there already exist artifact's with the same name, we will error.
+        num_outputs:
+            Must be consistent with `candidate_artifact_names`, if set. The number of
+            output artifacts to expected.
+        only_resolve_op_name:
+            If set, skip the artifact naming validation step. An empty list will be
+            returned for the artifact names in that case.
+    Returns:
+        A tuple of the operator name and artifact name(s), which the caller can use
+        to safely attach elements to the DAG.
+    """
+    # First, resolve the candidate op name. If we can overwrite an existing op name, then
+    # don't even bother finding an unallocated op name.
+    if not overwrite_existing_op_name:
+        prefix = candidate_op_name
+        curr_suffix = 1
+        while dag.get_operator(with_name=candidate_op_name) is not None:
+            candidate_op_name = prefix + " (%d)" % curr_suffix
+            curr_suffix += 1
+
+    # If we're overwriting an existing operator, temporarily delete the operator from the dag
+    # so that downstream names won't collide with our future naming checks.
+    else:
+        colliding_op = dag.get_operator(with_name=candidate_op_name)
+        if colliding_op is not None:
+            dag = apply_deltas_to_dag(
+                dag,
+                deltas=[
+                    RemoveOperatorDelta(op_id=colliding_op.id),
+                ],
+                make_copy=True,
+            )
+    op_name = candidate_op_name
+    if only_resolve_op_name:
+        return op_name, []
+
+    # Second, validate the artifact name(s) does not collide with other artifacts.
+    if candidate_artifact_names is None:
+        candidate_artifact_names = _construct_default_output_artifact_names_from_op(
+            op_name, num_outputs
+        )
+    assert candidate_artifact_names is not None
+
+    if isinstance(candidate_artifact_names, str):
+        candidate_artifact_names = [candidate_artifact_names]
+    assert isinstance(candidate_artifact_names, list)
+    assert num_outputs == len(candidate_artifact_names)
+
+    for candidate_artifact_name in candidate_artifact_names:
+        dag.validate_artifact_name(candidate_artifact_name)
+    artifact_names = candidate_artifact_names
+
+    return op_name, artifact_names
```

Comment:
Can we just return `candidate_artifact_names`?

Reply:
I wanted to make it clear at what point a candidate name became the resolved name though - it makes a little more sense for the op_name assignment above.